### DISCUSSION TITLE
JSON functions in SQL?

### DISCUSSION THREAD

**Discussion by ghost** [openobserve/openobserve#3668] [Q&A]
As a consequence of implementing the [flattening](https://github.com/openobserve/openobserve/issues/3216) deeply nested log data but also the use of serialized JSON arrays, I have trouble finding a way to make use of that data (again).

I was not able to find anything in the [SQL functions reference](https://openobserve.ai/docs/functions/) docs, nor in the [Apache DataFusion SQL reference](https://datafusion.apache.org/user-guide/sql/index.html) in regard to JSON data handling.

What I essentially want to achieve is to maintain a high ingestion efficiency (high throughput potential, save CPU time in general, ready/enough headroom for bursts) and "offload" the deeply nested data structure analysis (which is application-specific anyway) to the actual analysis phase (SQL queries).

Major use cases that I have at the moment:
1. in the `WHERE` clause, filter for log entries that do have a specific value inside an array value
2. in the `SELECT` clause, extract a specific field/path from a deeply nested JSON object to group on

Is this possible somehow?

P.S.: coming from a PostgreSQL background, so expecting to have something here as well ;-)

**Comment by prabhatsharma** [comment]
Not yet. We will be building a json extractor soon. However you should prioritize doing things at ingestion time in order to improve query performance with lower resources. In most cases that would be right decision. But most cases is not all cases, so make your decision.

**Reply by ghost** [reply]
OK, is there any way I can avoid the flattening only for specific messages and/or streams? That would be needed for our situation, since the impact of increasing the level of depth generally through `ZO_INGEST_FLATTEN_LEVEL` is quite high.

In general, the flattening is really good, since some external tools deployed in our environments produce really awkward JSON logs, but for some of our apps we really need to dig into specific fields for very specific analysis purposes (which are OK to be a bit slower in execution).

For the array case, I don't see any option at the moment, or did I miss something?

Or do you recommend transforming the array into a String→Boolean object/map at ingestion time? E.g.:
```
parsed_entity_tags: ["test", "public"]
```
...would become:
```
parsed_entity_tags_test: true
parsed_entity_tags_public: true
```
...which then could easily lead to an explosion in fields (especially if array values are generated/dynamic and not static), again harming performance.

Any other suggestion/recommendation?

**Reply by prabhatsharma** [reply]
JSON parse in SQL should be available soon.

**Reply by deviousasti** [reply]
Any updates on this?

**Comment by Manazsharma** [comment]
Hi there,  you can use [spath](https://openobserve.ai/docs/sql_reference/#spath) to extract fields from nested JSON. Also, O2 supports [datafusion-functions-json](https://github.com/datafusion-contrib/datafusion-functions-json) for array operations in SQL queries.

And to address your use-case, as to extract a specific field/path from a deeply nested JSON object, use [spath](https://openobserve.ai/docs/sql_reference/#spath)
ex: For `field_name` like {"nested": {"value": 23}}, extract 23 as nested_value, we need to add a where clause to do a full text search on the key here:
```sql
SELECT *,spath(field_name, 'nested.value') AS nested_value FROM "default" where field_name like '%nested%'
```
Lastly, to filter rows, you might have to stick to matching your JSON array values (e.g., 20 in {"items": [10, 20, 30]}) using full-text search with LIKE, as spath is limited to extracting specific paths and may not reliably filter values within arrays.
